src/passes/Inlining.cpp

```diff
@@ -170,26 +183,58 @@ static Expression*
 doInlining(Module* module, Function* into, InliningAction& action) {
   Function* from = action.contents;
   auto* call = (*action.callSite)->cast<Call>();
+  // Works for return_call, too
+  Type retType = module->getFunction(call->target)->result;
   Builder builder(*module);
-  auto* block = Builder(*module).makeBlock();
+  auto* block = builder.makeBlock();
   block->name = Name(std::string("__inlined_func$") + from->name.str);
-  *action.callSite = block;
+  if (call->isReturn) {
+    if (isConcreteType(retType)) {
+      *action.callSite = builder.makeReturn(block);
+    } else {
+      *action.callSite = builder.makeSequence(block, builder.makeReturn());
+    }
+  } else {
+    *action.callSite = block;
+  }
   // Prepare to update the inlined code's locals and other things.
   struct Updater : public PostWalker<Updater> {
+    Module* module;
     std::map<Index, Index> localMapping;
     Name returnName;
     Builder* builder;
-
     void visitReturn(Return* curr) {
       replaceCurrent(builder->makeBreak(returnName, curr->value));
     }
+    void replaceReturnCall(Expression* curr) {
+      if (isConcreteType(curr->type)) {
+        replaceCurrent(builder->makeBreak(returnName, curr));
+      } else {
+        replaceCurrent(builder->blockify(curr, builder->makeBreak(returnName)));
+      }
+    }
+    void visitCall(Call* curr) {
+      if (curr->isReturn) {
+        curr->isReturn = false;
```

Comment:
this changes the semantics, doesn't it? or can we never try to inline in a case where it would be noticeable (like looping using return calls)?

if it's safe then a comment why would be good.

Reply:
Nope, return_calls becomes more of a break_call so instead of returning from the new caller it just breaks out of the inlined function block. The inlining has no effect on stack growth because the tail called function reuses its caller's stack frame whether or not it is inlined (of course the frame can be bigger after inlining, but that seems ok). I'll add a comment to that effect.

return_call loops should be protected from unbounded inlining by the same rules governing inlining with normal recursion, so I don't think they're special. Is there something I'm missing there?